README.md
# EGG
The Empirical Galaxy Generator (EGG) is a set of tools to generate fake galaxy catalogs and images with realistic positions, morphologies and fluxes from the far-ultraviolet to the far-infrared. The catalogs are generated by egg-gencat, and stored in binary FITS tables (column oriented). Another program, egg-2skymaker, is used to convert the generated catalog into ASCII tables suitable for ingestion by [SkyMaker] to produce realistic high resolution images (e.g., Hubble-like), while egg-gennoise and egg-genmap can be used to generate the low resolution images (e.g., Herschel-like).

These tools can be used to test source extraction codes, or to evaluate the reliability of any map-based science (stacking, dropout identification, ...).

# Installation
You must have the [vif] library installed on your machine to compile EGG, configured at least with cfitsio and WCSlib support. [CMake] is used as a build system to handle dependency checks and compilation in a cross-platform way.

One you have installed all the dependencies, create yourself a directory called 'build' within the current directory. Navigate to the 'build' directory with your terminal, and call:

    cmake ../
    make
    # then
    make install
    # or
    sudo make install

The binaries will be generated in your system default binary folder. More detailed information can be found in the script [install.sh] or in the documentation ('doc/' folder).

# Usage
See the output of 'egg-gencat help' after installing the programs.
For further and more detailed help, see the documentation in the 'doc/' folder.

[SkyMaker]: http://www.astromatic.net/software/skymaker
[vif]: http://cschreib.github.io/vif/
[CMake]: http://www.cmake.org/
[install.sh]: https://raw.githubusercontent.com/cschreib/egg/master/doc/scripts/install.sh


CHANGELOG
v1.4.0
======

New/updated feature:
 - Added a limited set of optical emission lines (Ly-alpha, [OII], [OIII], Balmer series, [NII]) and far-infrared emission lines ([CII]157, [NII]205, CI, CO ladder). The line fluxes are tied to  the generated SFR, metallicity (through the FMR), LIR, or Mgas (through Mdust), with recipes calibrated on various observations, and the usual random scatter. Optical lines will be attenuated using a different attenuation than the continuum. The lines can be disabled by setting the flag 'no_nebular'.


v1.3.1
======

Bug fixes:
 - Fixed position angle being limited to -45 to +45 degrees instead of -90 to +90 degrees.


v1.3.0
======

New/updated feature:
 - A new IGM absorption method is added, using the Inoue+14 prescription and allowing for scatter in the IGM absorption at a given redshift due to varying lines of sight through the IGM. For this occasion, the "naive_igm" option has been removed: instead the user can now choose which IGM recipe they want to use using the "igm=..." option. See help for more info. The new Inoue+14 IGM with scatter is now the default in EGG. Other methods ("constant", equivalent to the previous "naive_igm=1", and "madau95", equivalent to the previous "naive_igm=0") do not have scatter implemented.
 - It is now possible to generate galaxies without any IGM absorption using "igm=none".


v1.2.1
======

New/updated feature:
 - Updated to latest vif API, no consequence at all on the code and outputs.


v1.2.0
======

New/updated feature:
 - The library of stellar SEDs was updated. The new version has several interesting improvements. First, each SED in the library is now garanteed to have its UVJ colors matching its position on the UVJ grid. Second, the SEDs were produced by fitting observed galaxies with a broader range of star formation histories, so they should display more diversity. Third, these fits were done excluding the broadbands containing [OIII] and Hbeta, therefore the contamination from emission lines is mitigated (this should enable adding emission lines to the EGG SEDs without biases).
 - The recipes for assigning UVJ colors to bulges and disks were revised to improve agreement with observations. The scatter of blue colors along and across the UVJ sequence is increased, so blue components at a fixed mass have a larger variety of SEDs. Quiescent galaxies can no longer have blue colors for either their disks or bulges; this reproduces better the bimodality in the UVJ diagram.

Bug fixes:
 - The stellar mass function at 0<z<0.5 was incorrectly interpolated (z=0 and z>=0.5 were not affected).


v1.1.0
======

New/updated feature:
 - Improved the treatment of IGM absorption. Now the absorption is computed on a per-galaxy basis, based on the redshift. Before that, the IGM absorption was quite poorly implemented, as a single value of the absorption for the entire catalog with no redshift dependence. The old behavior of the code can be recovered by setting the "naive_igm" flag on the command line.


v1.0.9
======

Bug fixes:
 - Fixed output SEDs produced by egg-getsed containing duplicate values when the component='...' option was not used. This fix should also improve performances of egg-gencat.


v1.0.8
======

Bug fixes:
 - Fixed a compilation error with recent versions of CFITSIO (CURL dependency).


v1.0.7
======

New/updated features:
 - Added Av as an output of the simulation.

Bug fixes:
 - Added sanity checks for input filters.


v1.0.6
======

New/updated features:
 - Added JWST and ALMA filters.

Changes in EGG recipes:
 - The mass functions can now reach z=15, however this is a wild extrapolation!
 - The recipes for Tdust and IR8 have been updated to the relations published in Schreiber et al. (2018).

Bug fixes:
 - Fixed documentation for 'filter_db' parameter.
 - Fixed issues with WCSLib and distortion in FITS headers.


v1.0.5
======

New/updated features:
 - Added the 'max_dz' option (default is 0.5) to avoid generating too broad redshift slices.

Bug fixes:
 - Fixed redshift bins getting too large when zmin is itself relatively large.
 - Fixed the 'min_dz' option doing nothing.
 - Stellar SEDs are no longer attributed to galaxies when the 'no_stellar' option is set.
 - Fixed columns being empty in the output FITS catalog, as some languages/libraries do not
   like this.


v1.0.4
======

New/updated features:
 - Added the rest-frame filters U, B and V (useful to compute UVJ colors).
 - 'egg-getsed' can now produce the full (disk+bulge) SED of a galaxy rather than only producing
   either the disk or the bulge SED. Simply omit the 'component' option to do that.
 - Changed the algorithm to determine the lowest mass generated as a function of redshift when the
   'maglim' option is used. This new algorithm takes into account the typical colors (and M/L) of
   galaxies as a function of mass and redshift, so that 'egg-gencat' generates fewer galaxies below
   the requested magnitude limit, while remaining complete above that limit.

Changes in EGG recipes:
 - The z=4 stellar mass function is now given a shallower faint end slope, and this slope is made
   also shallower at higher redshifts, with a cap at -2.1 to prevent exploding the counts.
 - The mass-to-light ratio of z>6 galaxies is adjusted to match the observations of Song+16.
 - Bluer templates are added to the stellar library to cover a wider region of the UVJ plane.
   This region is populated at high redshift and low masses.

Bug fixes:
 - Fixed linking error with libdwarf and libelf.
 - Fixed typo in libelf CMake script.
 - Fixed phy++ CMake script to account for new header structure.
 - Fixed typo in documentation.


v1.0.3
======

Changes in EGG recipes:
 - Tweaked the stellar mass functions at z>4 to match the observations of Grazian+14.
   This involves having a steeper low mass slope, and a re-calibrated evolution of the
   total stellar mass density at z>6.
 - Stopped the evolution of the UVJ colors beyond z=3.3 to prevent galaxies from becoming

ifni-2skymaker.cpp
#include <phypp.hpp>

void print_help();

// SkyMaker pixel type (see PIXTYPE in skymaker/src/fits/fitscat.h)
using SkyPixType = float;

int main(int argc, char* argv[]) {
    if (argc < 2) {
        print_help();
        return 1;
    }

    // Read command line arguments
    std::string band;
    std::string out_file;
    std::string img_dir;
    std::string template_file;
    double maglim = dnan;
    double inset = 0.0;
    bool verbose = false;
    float size_cap = fnan;
    std::string save_pixpos;

    read_args(argc-1, argv+1, arg_list(
        band, name(out_file, "out"), img_dir, maglim, size_cap, save_pixpos, verbose,
        name(template_file, "template")
    ));

    if (!img_dir.empty()) {
        img_dir = file::directorize(img_dir);
    }

    // Check for missing mandatory parameters
    bool bad = false;
    if (band.empty()) {
        error("missing band ID (band=...)");
        bad = true;
    }
    if (template_file.empty()) {
        error("missing SkyMaker template configuration file (template=...)");
        bad = true;
    }

    if (bad) return 1;

    // Read input catalog
    if (verbose) {
        note("reading catalog...");
    }

    struct {
        vec1u id;
        vec1d ra, dec;

        vec1f disk_angle,  disk_radius,  disk_ratio;
        vec1f bulge_angle, bulge_radius, bulge_ratio;

        vec2f flux_disk, flux_bulge;
        vec1s bands;
        vec1f lambda;
    } cat;

    fits::read_table(argv[1], ftable(
        cat.id, cat.ra, cat.dec,
        cat.disk_angle, cat.disk_radius, cat.disk_ratio,
        cat.bulge_angle, cat.bulge_radius, cat.bulge_ratio,
        cat.flux_disk, cat.flux_bulge, cat.bands, cat.lambda
    ));

    if (verbose) {
        note("found ", cat.id.size(), " galaxies");
    }

    vec1u idb = where(cat.bands == band);
    if (idb.empty()) {
        error("no band named '", band, "' in this catalog");
        return 1;
    } else if (idb.size() > 1) {
        error("multiple bands named '", band, "' in this catalog");
        return 1;
    }

    uint_t b = idb[0];

    // Read template SkyMaker configuration file
    double aspix = fnan;
    vec1s sky_param, sky_value; {
        std::ifstream file(template_file);

        uint_t l = 0;
        std::string line;
        while (std::getline(file, line)) {
            ++l;

            line = trim(line);
            if (line.empty() || line[0] == '#') continue;

            auto pos = line.find_first_of(" \t");
            if (pos == line.npos) {
                note("reading ", template_file, ":", l);
                error("ill formed line, expected \"PARAMETER   VALUE\"");
                return 1;
            }

            std::string param = line.substr(0, pos);
            std::string value = trim(line.substr(pos));

            sky_param.push_back(param);
            sky_value.push_back(value);

            if (param == "PIXEL_SIZE" && !from_string(value, aspix)) {
                note("reading ", template_file, ":", l);
                error("could not read pixel size value '", value, "'");
                return 1;
            }
        }
    }

    if (!is_finite(aspix)) {
        error("missing 'PIXEL_SIZE' parameter from SkyMaker configuration file");
        return 1;
    }

    // If missing, add some parameters we can provide
    if (count(sky_param == "LISTCOORD_TYPE") == 0) {
        sky_param.push_back("LISTCOORD_TYPE");
        sky_value.push_back("PIXEL");
    }

    if (count(sky_param == "WAVELENGTH") == 0) {
        sky_param.push_back("WAVELENGTH");
        sky_value.push_back(strn(cat.lambda[b]));

    }

    if (count(sky_param == "IMAGE_TYPE") == 0) {
        sky_param.push_back("IMAGE_TYPE");
        sky_value.push_back("SKY");
    }

    if (out_file.empty()) {
        out_file = file::remove_extension(argv[1])+"-"+band+".cat";
    }

    // Now start the real job
    file::mkdir(file::get_directory(out_file));

    // Build the quantities expected by SkyMaker
    if (verbose) {
        note("convert parameters for ingestion by SkyMaker...");
    }

    vec1f mag = uJy2mag(cat.flux_disk(_,b)+cat.flux_bulge(_,b));
    vec1f bt = cat.flux_bulge(_,b)/(cat.flux_disk(_,b)+cat.flux_bulge(_,b));

    // Convert effective radius to scale length, which is what SkyMaker expects
    // (NB: only for the disk, the bulge size has to be the effective radius)
    cat.disk_radius /= 1.678;

    // Try first to make a single, big image
    if (verbose) {
        note("define image size and pixel coordinates...");
    }

    // Roughly estimate galaxy size to check for border effects [pixels]
    vec1d gsize = 10*max(cat.disk_radius*1.678, cat.bulge_radius)/aspix;

    // Try a first naive projection where the first pixel is defined by the
    // most extreme galaxy in the corner of the field
    fits::make_wcs_header_params wcs_params;
    wcs_params.pixel_scale = aspix;
    wcs_params.pixel_ref_x = 1.0; wcs_params.pixel_ref_y = 1.0;
    wcs_params.sky_ref_ra = max(cat.ra); wcs_params.sky_ref_dec = min(cat.dec);

    fits::header hdr;
    if (!fits::make_wcs_header(wcs_params, hdr)) {
        error("could not make WCS header");
        return 1;
    }

    vec1d x, y;
    fits::wcs wcs_big(hdr);
    fits::ad2xy(wcs_big, cat.ra, cat.dec, x, y);

    // Now, because of the spherical projection, some galaxies can have negative pixel
    // coordinates even though we picked the most extreme one in RA and Dec.
    // Figure out how negative this can be and adjust the reference pixel accordingly.
    // Also add the inset that was requested by the user, and take into account the
    // extent of each galaxy
    double dx = inset/aspix - (min(x - gsize) - 1);
    double dy = inset/aspix - (min(y - gsize) - 1);
    wcs_params.pixel_ref_x += dx;
    wcs_params.pixel_ref_y += dy;

    if (!fits::make_wcs_header(wcs_params, hdr)) {
        error("could not make WCS header");
        return 1;
    }

    wcs_big = fits::wcs(hdr);
    fits::ad2xy(wcs_big, cat.ra, cat.dec, x, y);

    // Compute expected full image size in GB
    uint_t nx = ceil(max(x + gsize) + inset/aspix);
    uint_t ny = ceil(max(y + gsize) + inset/aspix);

    // Select only the galaxies with valid of bright enough fluxes
    vec1u ids;
    if (is_finite(maglim)) {
        ids = where(mag < maglim);
        if (ids.empty()) {
            error("no source brighter than ", maglim);
            note("minimum magnitude is ", min(mag));
            return 1;
        }

        if (verbose) {
            note("selecting ", ids.size(), "/", cat.id.size(), " galaxies");
        }
    } else {
        ids = uindgen(x.size());
        vec1u idbad = where(!is_finite(mag));
        mag[idbad] = 99;
    }

    double img_size = double(nx)*ny*sizeof(SkyPixType)/pow(1024.0, 3);

    // Inline function to write a catalog to disk
    auto write_catalog = [&](std::string ofile, const vec1u& oids, uint_t sx, uint_t sy){
        // Write catalog
        file::write_table_hdr(ofile, 16,
            {"type", "x", "y", "mag", "bt",
            "bulge_radius", "bulge_ratio", "bulge_angle",
            "disk_radius", "disk_ratio", "disk_angle"},
            replicate(200u, oids.size()), x[oids], y[oids],
            mag[oids], bt[oids],
            cat.bulge_radius[oids], cat.bulge_ratio[oids], cat.bulge_angle[oids],
            cat.disk_radius[oids], cat.disk_ratio[oids], cat.disk_angle[oids]
        );

        std::string obase = file::remove_extension(ofile);

        // Write FITS header
        std::string hdr_file = obase+"-hdr.txt";
        std::ofstream ohdr(hdr_file);
        vec1s shdr = cut(hdr, 80);
        for (auto& s : shdr) {
            ohdr << s << "\n";
        }

        // Write SkyMaker configuration file
        std::ofstream oconf(obase+"-sky.conf");
        std::string img_name = file::get_basename(obase);

        // Copy the parameters from the template
        vec1s tpar = sky_param;
        vec1s tval = sky_value;

        // Add our own
        tpar.push_back("IMAGE_NAME");
        tval.push_back(img_dir+img_name+"-sci.fits");
        tpar.push_back("IMAGE_SIZE");
        tval.push_back(strn(sx)+","+strn(sy));
        tpar.push_back("IMAGE_HEADER");
        tval.push_back(hdr_file);

        // Write to file
        tpar = align_left(tpar, max(length(tpar)));
        for (uint_t i : range(tpar)) {
            oconf << tpar[i] << " " << tval[i] << "\n";
        }
    };

    vec1u tilex(x.size());
    vec1u tiley(x.size());

    if (is_finite(size_cap) && img_size > size_cap) {
        // The whole image would be too big, we need to split it into smaller sections
        // of equal size.
        if (verbose) {
            note("full image would be ", nx, " x ", ny, " (", img_size, " GB)");
        }

        // Estimate the number of such sections
        const double fudge = 0.95; // make sure we are below the maximum size
        double nsec = 1.0/(fudge*size_cap/img_size);

        // Estimate the number of X and Y sections
        double ar = nx/double(ny);
        double tnsx = sqrt(ar*nsec);
        double tnsy = nsec/tnsx;

        uint_t nsx, nsy;
        if (fabs(round(tnsx) - tnsx) < fabs(round(tnsy) - tnsy)) {
            nsx = round(tnsx);
            nsy = ceil(nsec/nsx);
        } else {
            nsy = round(tnsy);
            nsx = ceil(nsec/nsy);
        }

        nx = ceil(nx/double(nsx));
        ny = ceil(ny/double(nsy));

        if (verbose) {
            double sec_size = double(nx)*ny*sizeof(SkyPixType)/pow(1024.0, 3);
            note("will require ", nsx, "x", nsy, " sections (", sec_size, " GB each)");
            note("section dimensions: ", nx, ",", ny);
        }

        if (verbose) {
            note("write catalogs...");
        }

        tilex.resize(x.size());
        tiley.resize(x.size());

        // Keep "old" coordinates from the whole image
        vec1d ox = x;
        vec1d oy = y;
        // And these will be the new coordinates of each galaxy
        vec1d tx = x;
        vec1d ty = y;

        auto pg = progress_start(nsx*nsy);
        for (uint_t iy : range(nsy))
        for (uint_t ix : range(nsx)) {
            // Pick galaxies that fall in this section
            // NB: pixel coordinates in FITS standard are in [1,N] and '1' is the center
            // of the first pixel (which therefore covers 0.5 to 1.5)
            vec1u idi = ids[where(ox[ids]+gsize[ids] >= nx*ix+0.5     &&
                                  ox[ids]-gsize[ids] <= nx*(ix+1)+0.5 &&
                                  oy[ids]+gsize[ids] >= ny*iy+0.5     &&
                                  oy[ids]-gsize[ids] <= ny*(iy+1)+0.5)];

            // This is the "strict" list, i.e., only galaxies whose center lies in this tile
            vec1u idis = where(ox[idi] >= nx*ix+0.5 && ox[idi] < nx*(ix+1)+0.5 &&
                               oy[idi] >= ny*iy+0.5 && oy[idi] < ny*(iy+1)+0.5);

            // Convert pixel coordinates to local frame
            wcs_params.pixel_ref_x = dx+1-nx*double(ix);
            wcs_params.pixel_ref_y = dy+1-ny*double(iy);

            if (!fits::make_wcs_header(wcs_params, hdr)) {
                error("could not make WCS header");
                return 1;
            }

            vec1d ttx, tty;
            fits::ad2xy(fits::wcs(hdr), cat.ra[idi], cat.dec[idi], ttx, tty);
            x[idi]        = ttx;       y[idi]        = tty;
            tx[idi[idis]] = ttx[idis]; ty[idi[idis]] = tty[idis];

            // Write the catalog for this section
            std::string six = align_right(strn(ix+1), strn(nsx).size(), '0');
            std::string siy = align_right(strn(iy+1), strn(nsy).size(), '0');

            auto spl = file::split_extension(out_file);
            write_catalog(spl.first+"-"+six+"-"+siy+spl.second, idi, nx, ny);

            tilex[idi[idis]] = ix+1;
            tiley[idi[idis]] = iy+1;

            if (verbose) progress(pg);
        }

        std::swap(tx, x);
        std::swap(ty, y);
    } else {
        // We will make a single big image
        if (verbose) {
            note("image dimensions: ", nx, ",", ny, " (", img_size, " GB)");
        }

        if (verbose) {
            note("write catalog...");
        }

        write_catalog(out_file, ids, nx, ny);

        tilex = replicate(1u, x.size());
        tiley = replicate(1u, x.size());
    }

    if (!save_pixpos.empty()) {
        note("write pixel positions...");
        file::mkdir(file::get_directory(save_pixpos));
        fits::write_table(save_pixpos, ftable(cat.id, x, y, tilex, tiley));
    }

    if (verbose) {
        note("done.");
    }

    // TODO: fix SkyMaker and border objects

    return 0;
}

void print_help() {
    using namespace format;

    auto argdoc = [](const std::string& name, const std::string& type,
        const std::string& desc) {

        std::string header = " - "+name+" "+type;
        print(header);

        std::string indent = "    ";
        vec1s w = wrap(indent+desc, 80, indent);

        for (auto& s : w) {
            print(s);
        }
    };

    print("ifni-2skymaker v1.0");
    print("usage: ifni-2skymaker cat.fits [parameters and options]\n");

    print("List of mandatory parameters (no default):");
    argdoc("band", "[string]", "name of the photometric band for which a SkyMaker "
        "input catalg will be created");
    argdoc("template", "[string]", "file containing a template SkyMaker configuration "
        "setup that this program will use to create the final configuration file. This "
        "template file must contain the 'PIXEL_SIZE' parameter. If missing, this program "
        "will add 'WAVELENGTH' and 'LISTCOORD_TYPE' (which has to be 'PIXELS').");

    print("");
    print("List of options:");
    argdoc("out", "[string]", "name of the output catalog (default: [catalog]-[band].cat)");
    argdoc("maglim", "[float]", "galaxies fainter than this magnitude will not be "
        "present in the output catalog (default: none)");
    argdoc("inset", "[double, arcsec]", "add an empty border around the input catalog "
        "to prevent galaxies from being truncated by the edge of the image "
        "(default: 5\")");
    argdoc("size_cap", "[float, GB]", "split the survey in multiple tiles of equal area "
        "that will each be rendered into a separate image, so that the size of each "
        "such image is less than 'size_cap' (expressed in GigaBytes, default: none). NB: "
        "SkyMaker requires the whole image to fit inside your RAM memory, so this value "
        "should be no larger than the total amount of RAM memory available on your "
        "computer.");
    argdoc("save_pixpos", "[string]", "specify the name of the file into which the pixel "
        "coordinates and the name of the image of each galaxy will be written (default: "
        "not saved). This is most useful when 'size_cap' is used, as this catalog "
        "provides and easy way to figure out in which of the tile each galaxy is located.");
    argdoc("verbose", "[flag]", "display information about the progress of the program "
        "in the terminal");

    print("");
}
